<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • platjam</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">platjam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.73.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="http://github.com/jmw86069/platjam/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="http://github.com/jmw86069/platjam/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.73.900" id="platjam-0073900">platjam 0.0.73.900<a class="anchor" aria-label="anchor" href="#platjam-0073900"></a></h2>
<ul><li>dependency version bump for colorjam (&gt;= 0.0.26.900)</li>
</ul><div class="section level3">
<h3 id="changes-to-existing-functions-0-0-73-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-73-900"></a></h3>
<ul><li><p><code><a href="../reference/quick_complement_color.html">quick_complement_color()</a></code> was updated to match recent updates to colorjam version 0.0.26.900. This change also affected <code><a href="../reference/design2colors.html">design2colors()</a></code></p></li>
<li>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code></p>
<ul><li>
<code>preset="dichromat2"</code> new default, so colors begin with yellow, more conducive for applying to control groups than starting with red.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.72.900" id="platjam-0072900">platjam 0.0.72.900<a class="anchor" aria-label="anchor" href="#platjam-0072900"></a></h2>
<div class="section level3">
<h3 id="migration-of-coverage-heatmap-functions-to-coverjam-0-0-72-900">migration of coverage heatmap functions to coverjam<a class="anchor" aria-label="anchor" href="#migration-of-coverage-heatmap-functions-to-coverjam-0-0-72-900"></a></h3>
<p>The functions related to coverage heatmaps are being migrated into <code>"jmw86069/coverjam"</code> Github-hosted R package, for cleaner maintenance of those functions.</p>
</div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.71.900" id="platjam-0071900">platjam 0.0.71.900<a class="anchor" aria-label="anchor" href="#platjam-0071900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-71-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-71-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_metabolomics_niehs.html">import_metabolomics_niehs()</a></code></p>
<ul><li>added optional step to import <code>"compounds_pos.txt"</code> and <code>"compounds_neg.txt"</code> files, which contains important annotations for which measurements are imputed by the upstream software, and using which impute method.</li>
<li>When the files above are not present, and the file <code>"1_DataProcessed.zip"</code> exists, it is used to extract the raw files as named above.</li>
<li>When no matching files are detected, this import step is skipped.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-71-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-71-900"></a></h3>
<ul><li>
<p><code><a href="../reference/convert_imputed_assays_to_na.html">convert_imputed_assays_to_na()</a></code></p>
<ul><li>utility function used to apply filtering based upon the impute flags, thereby converting imputed data matching the relevant flags to <code>NA</code>, and storing in a new <code>assays()</code> entry for downstream analysis.</li>
</ul></li>
<li>
<p><code><a href="../reference/process_metab_compounds_file.html">process_metab_compounds_file()</a></code></p>
<ul><li>Internal function used to import full compounds data supplied as a <code>data.frame</code> after loading from the <code>"1_DataProcessed.zip"</code> file.</li>
<li>The function will merge data into a <code>se_list</code> object if supplied, otherwise it will return the full <code>se</code> object including each assay name imported from the file.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.70.900" id="platjam-0070900">platjam 0.0.70.900<a class="anchor" aria-label="anchor" href="#platjam-0070900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-70-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-70-900"></a></h3>
<ul><li>
<p>Added missing package prefix for several calls to <code><a href="https://rdrr.io/pkg/jamba/man/nameVector.html" class="external-link">jamba::nameVector()</a></code>, or <code><a href="https://rdrr.io/pkg/jamba/man/printDebug.html" class="external-link">jamba::printDebug()</a></code>. Affected functions:</p>
<ul><li><code><a href="../reference/import_nanostring_rcc.html">import_nanostring_rcc()</a></code></li>
<li><code><a href="../reference/coverage_matrix2nmat.html">coverage_matrix2nmat()</a></code></li>
<li><code><a href="../reference/frequency_matrix2nmat.html">frequency_matrix2nmat()</a></code></li>
<li><code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.69.900" id="platjam-0069900">platjam 0.0.69.900<a class="anchor" aria-label="anchor" href="#platjam-0069900"></a></h2>
<ul><li>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code></p>
<ul><li>input recognizes <code>DataFrame</code> which is commonly produced from <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::colData()</a></code>.</li>
<li>input recognizes <code>tbl_df</code> (tibble) and <code>matrix</code> by coercing to <code>data.frame</code>.</li>
<li>column sort can be reversed with “-” prefix. For example <code>"-Treatment"</code> will reverse the sort order in column <code>"Treatment"</code>.</li>
<li>allow empty <code>group_colnames</code> so colors are assigned in all columns</li>
<li>tolerate various combinations of <code>group_colnames</code>, <code>class_colnames</code> with non-ideal cardinality, by using different workarounds strategies</li>
<li>new arguments <code>Crange</code>,<code>Lrange</code> which are passed to <code><a href="https://rdrr.io/pkg/colorjam/man/rainbowJam.html" class="external-link">colorjam::rainbowJam()</a></code> to set a less saturated color palette by default.</li>
</ul></li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.68.900" id="platjam-0068900">platjam 0.0.68.900<a class="anchor" aria-label="anchor" href="#platjam-0068900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-68-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-68-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_metabolomics_niehs.html">import_metabolomics_niehs()</a></code></p>
<ul><li>Fixed bug when importing files that may contain duplicated rows in metadata, and in each data matrix file. The issue appears to be pure duplication of rows, so the workaround is to print a warning, and retain non-duplicated records per SampleID value.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.67.900" id="platjam-0067900">platjam 0.0.67.900<a class="anchor" aria-label="anchor" href="#platjam-0067900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-67-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-67-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_metabolomics_niehs.html">import_metabolomics_niehs()</a></code></p>
<ul><li>imports metabolomics data specifically in file formats used by the NIEHS Metabolomics Core Facility. The output is a <code>list</code> of <code>SummarizedExperiment</code> objects, typically including positive and negative ionization data.</li>
</ul></li>
<li>
<p><code><a href="../reference/curate_se_colData.html">curate_se_colData()</a></code></p>
<ul><li>wrapper function to <code><a href="../reference/curate_to_df_by_pattern.html">curate_to_df_by_pattern()</a></code> specifically for <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::colData()</a></code>, updating data in place.</li>
<li>optionally this step can subset columns for those that match curated patterns.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.66.900" id="platjam-0066900">platjam 0.0.66.900<a class="anchor" aria-label="anchor" href="#platjam-0066900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-66-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-66-900"></a></h3>
<ul><li>
<code><a href="../reference/import_salmon_quant.html">import_salmon_quant()</a></code> was using the older version of <code><a href="../reference/curate_to_df_by_pattern.html">curate_to_df_by_pattern()</a></code> supplied by <code>slicejam</code> instead of the slightly updated version in <code>platjam</code> (this package). Not such a big issue except that <code>slicejam</code> is not yet public.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.65.900" id="platjam-0065900">platjam 0.0.65.900<a class="anchor" aria-label="anchor" href="#platjam-0065900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-65-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-65-900"></a></h3>
<ul><li>added genejam Github information to DESCRIPTION, previously this package was not properly referenced as part of Github and would cause the package dependency to fail.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.64.900" id="platjam-0064900">platjam 0.0.64.900<a class="anchor" aria-label="anchor" href="#platjam-0064900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-64-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-64-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_nanostring_csv.html">import_nanostring_csv()</a></code></p>
<ul><li>new argument <code>curation_txt</code> for optional sample annotations during import; behaves exactly as with <code><a href="../reference/import_proteomics_mascot.html">import_proteomics_mascot()</a></code>.</li>
<li>TODO: will add this argument to <code><a href="../reference/import_nanostring_rcc.html">import_nanostring_rcc()</a></code>
</li>
<li>TODO: will create generic function for curation into SE <code>colData()</code>
</li>
<li>fixed bug not prefixing <code>rowData()</code>
</li>
</ul></li>
<li>
<p><code><a href="../reference/import_proteomics_mascot.html">import_proteomics_mascot()</a></code></p>
<ul><li>updated to fix error when colnames are changed by <code>curation_txt</code>
</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.63.900" id="platjam-0063900">platjam 0.0.63.900<a class="anchor" aria-label="anchor" href="#platjam-0063900"></a></h2>
<ul><li>added testthis unit tests for some functions</li>
</ul><div class="section level3">
<h3 id="changes-to-existing-functions-0-0-63-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-63-900"></a></h3>
<ul><li>
<p><code><a href="../reference/assign_numeric_colors.html">assign_numeric_colors()</a></code></p>
<ul><li>new argument <code>color_max</code> to define a fixed upper <code>numeric</code> threshold for the maximum color in the color gradient.</li>
</ul></li>
<li>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code></p>
<ul><li>Changes to enable <code>color_sub</code> to accept color ramp names in addition to accepting single colors.</li>
<li>Slightly cleaned up some logic during color assignment.</li>
<li>New argument <code>color_sub_max</code> intended to be passed to <code><a href="../reference/assign_numeric_colors.html">assign_numeric_colors()</a></code> argument <code>color_max</code> for particular column names.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.62.900" id="platjam-0062900">platjam 0.0.62.900<a class="anchor" aria-label="anchor" href="#platjam-0062900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-62-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-62-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_proteomics_mascot.html">import_proteomics_mascot()</a></code></p>
<ul><li>Added more robust detection of changes in column order between the input Excel file and the curation_txt data.</li>
<li>The output column order now matches the order of the curation_txt file, which should be helpful when defining the control group in statistical contrasts.</li>
</ul></li>
<li>
<p><code><a href="../reference/curate_to_df_by_pattern.html">curate_to_df_by_pattern()</a></code></p>
<ul><li>new argument <code>order_priority="df"</code> to indicate the order of rows in the <code>data.frame</code> returned will match the curation <code>df</code>.</li>
<li>option <code>order_priority="x"</code> will order rows in the <code>data.frame</code> returned by the order of entries in <code>x</code>.</li>
</ul></li>
<li>
<p><code><a href="../reference/import_lipotype_csv.html">import_lipotype_csv()</a></code></p>
<ul><li>minor changes to ensure the order of curated rows and sample identifiers were matched when the curation data did not contain <code>"Filename"</code> column, by default uses the last column returned by <code><a href="../reference/curate_to_df_by_pattern.html">curate_to_df_by_pattern()</a></code>.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.61.900" id="platjam-0061900">platjam 0.0.61.900<a class="anchor" aria-label="anchor" href="#platjam-0061900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-61-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-61-900"></a></h3>
<p>Proteomics data is sometimes provided in separate results files, which can later be merged into one larger format for downstream analysis.</p>
<ul><li>
<p><code><a href="../reference/merge_proteomics_se.html">merge_proteomics_se()</a></code></p>
<ul><li>This function provides a specific method of merging two <code>SummarizedExperiment</code> objects, with heuristics appropriate for proteomics expression data.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.60.900" id="platjam-0060900">platjam 0.0.60.900<a class="anchor" aria-label="anchor" href="#platjam-0060900"></a></h2>
<ul><li>
<code>jamba</code> dependency was bumped to 0.0.89.900</li>
<li>added <code>SummarizedExperiment</code> and <code>XML</code> to dependencies</li>
</ul><div class="section level3">
<h3 id="updates-to-existing-functions-0-0-60-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-60-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_nanostring_csv.html">import_nanostring_csv()</a></code> and <code><a href="../reference/import_nanostring_rcc.html">import_nanostring_rcc()</a></code></p>
<ul><li>both functions now return data transformed with <code>log2(1 + x)</code>
</li>
<li>CSV import will name the assay <code>"norm"</code> if the CSV file contains that substring, otherwise defaults to <code>"raw"</code>, consistent with RCC import.</li>
<li>The <code>assay_name</code> can be defined upfront for both functions.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-in-development-0-0-60-900">new functions in development<a class="anchor" aria-label="anchor" href="#new-functions-in-development-0-0-60-900"></a></h3>
<ul><li>
<p><code><a href="../reference/validate_heatmap_params.html">validate_heatmap_params()</a></code></p>
<ul><li>This function begins a process of making the coverage heatmaps into a command-line tool. Work in progress.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.59.900" id="platjam-0059900">platjam 0.0.59.900<a class="anchor" aria-label="anchor" href="#platjam-0059900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-59-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-59-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_salmon_quant.html">import_salmon_quant()</a></code></p>
<ul><li>new arguments <code>curate_tx_from</code>, <code>curate_tx_to</code> to allow custom curation of transcript identifiers, specifically for transcripts where the FASTA workflow introduces <code>"(-)"</code> and <code>"(+)"</code> to the identifiers. This situation occurred when adding gene body transcripts to the Thuman T2Tv2.0 genome GTF entries.</li>
</ul></li>
<li>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code></p>
<ul><li>argument <code>color_sub</code> will convert color names to hex format, for compatibility with some downstream uses, for example <code><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">knitr::kable()</a></code> and <code><a href="https://rdrr.io/pkg/jamba/man/kable_coloring.html" class="external-link">jamba::kable_coloring()</a></code> when used in an Rmarkdown document.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.58.900" id="platjam-0058900">platjam 0.0.58.900<a class="anchor" aria-label="anchor" href="#platjam-0058900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-58-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-58-900"></a></h3>
<ul><li>
<code><a href="../reference/import_salmon_quant.html">import_salmon_quant()</a></code> was errantly checking the class of <code>tx2gene</code>, causing it to ignore user-supplied <code>tx2gene</code>. This issue affected import when <code>gtf</code> was not supplied, which was much more common.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-58-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-58-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_salmon_quant.html">import_salmon_quant()</a></code></p>
<ul><li><p>the list of assays now use names directly from <code>tximport()</code>, specifically <code>c("counts", "abundance", "length")</code>, and not <code>c("counts", "tpm")</code>.</p></li>
<li><p><code>TxSE</code> now contains assayName <code>"length"</code>, and metadata item <code>"countsFromAbundance"</code>, so the output can be used with <code><a href="https://rdrr.io/pkg/tximport/man/summarizeToGene.html" class="external-link">tximport::summarizeToGene()</a></code> as a second step outside this import process.</p></li>
<li>
<p>new argument option <code>import_types="gene_body"</code> designed specifically when the GTF or <code>tx2gene</code> contain entries representing unspliced multi-exon genes, in addition to spliced proper transcripts. When present:</p>
<ul><li>
<code>import_types="gene"</code> will summarize transcripts to gene level, excluding entries annotated <code>"gene_body"</code>, returned as <code>GeneSE</code>.</li>
<li>
<code>import_types="gene_body"</code> will summarize transcripts to gene level, including entries annotated <code>"gene_body"</code>, returned as <code>GeneBodySE</code>.</li>
<li>
<code>import_types="gene_tx"</code> will summarize transcripts to gene level, and separately represents <code>"gene_body"</code> entries, returned as <code>GeneTxSE</code>.</li>
</ul></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.57.900" id="platjam-0057900">platjam 0.0.57.900<a class="anchor" aria-label="anchor" href="#platjam-0057900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-57-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-57-900"></a></h3>
<ul><li>
<code><a href="../reference/import_nanostring_csv.html">import_nanostring_csv()</a></code> is a basic import function for the common CSV style export from Nanostring software. This version has minimal extra capabilities, and produces a simple <code>SummarizedExperiment</code> object.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.56.900" id="platjam-0056900">platjam 0.0.56.900<a class="anchor" aria-label="anchor" href="#platjam-0056900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-56-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-56-900"></a></h3>
<ul><li>
<code><a href="../reference/import_proteomics_PD.html">import_proteomics_PD()</a></code> was updated to fix an errant check of the protein colnames. This check occurs to verify that the accession colnames did not contain delimited values, which causes the colnames to be split into individual values by <code><a href="https://rdrr.io/pkg/jamba/man/makeNames.html" class="external-link">jamba::makeNames()</a></code> with suffix. When that happens, the original colname is added before the split columns so the original data can be verified.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-56-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-56-900"></a></h3>
<ul><li>
<code><a href="../reference/import_proteomics_mascot.html">import_proteomics_mascot()</a></code> is a version 1 importer for MASCOT and similar data, which expects protein-level data with <code>totalIntensity</code> and <code>numSpectra</code> columns for each sample.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.55.900" id="platjam-0055900">platjam 0.0.55.900<a class="anchor" aria-label="anchor" href="#platjam-0055900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-55-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-55-900"></a></h3>
<ul><li>
<code><a href="../reference/import_proteomics_PD.html">import_proteomics_PD()</a></code> was updated to remove dependency on <code>slicejam</code> R package which is currently private during internal development. The only dependent function <code><a href="../reference/curate_to_df_by_pattern.html">curate_to_df_by_pattern()</a></code> was moved into this package.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.54.900" id="platjam-0054900">platjam 0.0.54.900<a class="anchor" aria-label="anchor" href="#platjam-0054900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-54-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-54-900"></a></h3>
<ul><li>
<code><a href="../reference/design2colors.html">design2colors()</a></code> caused an error when there were no leftover categorical colors to be assigned after the initial group/class/lightness, then user-defined colors by colname. This bug has been fixed.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.53.900" id="platjam-0053900">platjam 0.0.53.900<a class="anchor" aria-label="anchor" href="#platjam-0053900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-53-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-53-900"></a></h3>
<ul><li>
<code><a href="../reference/import_nanostring_rlf.html">import_nanostring_rlf()</a></code> is an import function to load Nanostring codeset data provided in the form of an RLF file.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-53-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-53-900"></a></h3>
<ul><li>
<code><a href="../reference/design2colors.html">design2colors()</a></code> now calls <code><a href="https://rdrr.io/pkg/colorjam/man/blend_colors.html" class="external-link">colorjam::blend_colors()</a></code> to define colors for each class. The method is still in development, so the specific methods may change. The output is intended to represent the group colors contained in each class, we’ll see how well it fares.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-53-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-53-900"></a></h3>
<ul><li>
<code><a href="../reference/design2colors.html">design2colors()</a></code> was updated to convert columns with class <code>"table"</code> into <code>"integer"</code>, otherwise weird things happened during coersion, causing an error.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.52.900" id="platjam-0052900">platjam 0.0.52.900<a class="anchor" aria-label="anchor" href="#platjam-0052900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-52-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-52-900"></a></h3>
<ul><li>
<code><a href="../reference/mean_hue.html">mean_hue()</a></code> is a simple function to return a mean hue in degrees.</li>
<li>
<code><a href="../reference/print_color_list.html">print_color_list()</a></code> is a convenience function to convert a color list that may contain color vectors, or color functions, into a list of named color vectors. It will print to the console using those colors.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-52-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-52-900"></a></h3>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code> was updated to to handle an edge condition.</p>
<ul><li>
<p>Refactoring the workflow:</p>
<ul><li>
<p>define colors by design:</p>
<ul><li>define colors to <code>group_colnames</code> optionally partitioned by <code>class_colname</code>
</li>
<li>define gradient colors to <code>lightness_colnames</code>
</li>
<li>define class colors using the mean hue for groups contained in each class</li>
</ul></li>
<li>
<p>Apply design colors to all matching columns:</p>
<ul><li>Test all columns for compatible cardinality with group, lightness.</li>
<li>Note: column names in <code>color_sub</code> will be ignored at this step, instead those columns use <code>color_sub</code> to generate a gradient.</li>
<li>Assign colors to each compatible column.</li>
<li>Note this step will include extra annotation columns that have consistent cardinality.</li>
<li>(This might be the coolest step tbh, since it figures out which columns match the design.)</li>
</ul></li>
<li>
<p>Expand <code>color_sub</code> with newly defined colors.</p>
<ul><li>Intent is to assign consistent categorical colors to values present in other columns.</li>
</ul></li>
<li>
<p>For columns without assigned colors:</p>
<ul><li>Apply <code>color_sub</code> to the column when it matches the column name.</li>
<li>Numeric columns produce a color function.</li>
<li>Character columns produce a named color vector.</li>
</ul></li>
<li>
<p>Assemble all remaining unassigned values:</p>
<ul><li>Numeric columns use the column name as its “value”, to define one color to the column header.</li>
<li>Remaining <code>character</code> values defined in <code>color_sub</code> are assigned.</li>
<li>Remaining unassigned <code>character</code> values, and <code>character</code> column names of numeric columns are assigned categorical colors.</li>
</ul></li>
<li>
<p>Apply new categorical colors</p>
<ul><li>For <code>numeric</code> columns, generate color function using the categorical color assigned to that column name.</li>
<li>For <code>character</code> columns, generate a color vector.</li>
</ul></li>
</ul></li>
<li><p>Note that the mechanism used may change, I am still testing how the assumptions work with real world data.</p></li>
<li><p>New behavior: When <code>color_sub</code> is assigned to a value in <code>colnames(x)</code>, the cardinality is not checked for consistency with group color assignments, and instead will use the color assigned for that column.</p></li>
<li>
<p>New argument <code>force_consistent_colors</code>:</p>
<ul><li><p>During color assignment, sometimes a column value is assigned one color in a column, and a different color in another column. It usually happens when <code>color_sub</code> defines a color for a column, where that column unintentionally shares a value with another column that gets colors assigned earlier in the chain.</p></li>
<li><p>New argument <code>force_consistent_colors=TRUE</code> forces values to be assigned only one color. It seems obvious to use this option, but why not always?</p></li>
<li><p>For purely text strings, like “Treated” and “Untreated” one would think any time “Treated” appears in a graphic, it should use the same color. Yes. However, sometimes the number “2” appears in a column, as character or integer, it doesn’t matter for this purpose. I have seen something like “sample_num” with values “1”, “2”, “3”, etc. Their cardinality matches 1:1 with class/group/lightness color assignments defined by this function, so these values are usually also assigned to the group colors. However, these values may appear in another column defined by <code>color_sub</code>, intended to apply a gradient of that color to values in the column. What if that column contained values <code>c(2, 20, 500, 1000)</code>? Only these values would be colorized: <code>c(20, 50, 1000)</code>, while <code>2</code> would be assigned the color from group colors. It looks weird.</p></li>
<li>
<p>There are workarounds:</p>
<ul><li>pre-filter columns which might have this effect. In the case above, remove <code>"sample_num"</code> before calling <code><a href="../reference/design2colors.html">design2colors()</a></code> so that column is not assigned a color.</li>
<li>associate all numeric columns with <code>color_sub</code>
</li>
</ul></li>
<li><p>The end result, which is intended, is to allow columns to have independent color assignments when values in those columns are not related. In the case above, the number <code>"2"</code> means something different in one column than another column.</p></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.51.900" id="platjam-0051900">platjam 0.0.51.900<a class="anchor" aria-label="anchor" href="#platjam-0051900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-51-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-51-900"></a></h3>
<ul><li>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code> was updated to adjust the low end of color gradient.</p>
<ul><li>specifically when the low-high numeric range is above zero, for example <code>c(20, 50)</code>. The adjustment lowers the low end of color scale 5% the diff, so the low actual color will not be white.</li>
<li>Also the function was using <code><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty()</a></code> to generate reasonable numeric steps across the numeric range, and was using those steps to define the color gradient. This method had weird effects, partly due to <code><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty()</a></code> making choices about the best visible number for display, instead of the point range. Instead, the color gradient is applied to the calculated number range, then that function is used to generate a color function across the <code><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty()</a></code> scale. I’m not sure it is ideal when <code><a href="https://rdrr.io/r/base/pretty.html" class="external-link">pretty()</a></code> does not color the numeric range, but it seems reasonable so far.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.50.900" id="platjam-0050900">platjam 0.0.50.900<a class="anchor" aria-label="anchor" href="#platjam-0050900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-50-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-50-900"></a></h3>
<ul><li>
<p><code>import_proteomics_pd()</code> was updated to handle absence of <code>curation_txt</code> where sample names and labels may or may not be renamed for each abundance matrix.</p>
<ul><li>Specifically <code>label_colname</code> is detected as <code>"Label"</code>, any column containing <code>"Label"</code>, then <code>"^Input$"</code>. Note <code>"Input"</code> is the default first colname produced when there is no <code>curation_txt</code> file, and should represent a unique column identifier.</li>
</ul></li>
<li>
<p><code><a href="../reference/design2colors.html">design2colors()</a></code> was updated:</p>
<ul><li>
<code>class_colnames</code> is required to have 1-to-many cardinality with <code>group_colnames</code>, or throws an error. As a result, class is guaranteed to be associated only with one set of columns.</li>
<li>class is assigned a color hue using the mean color hue of the groups it contains.</li>
<li>
<code>color_sub</code> will apply a gradient color to <code>numeric</code> column values.</li>
<li>TODO: add some ability to pre-define class color with <code>color_sub</code>
</li>
<li>TODO: add some ability to define numeric colors or color function for columns that contain <code>numeric</code> data. They are currently handled as categorical values, which is not ideal.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.49.900" id="platjam-0049900">platjam 0.0.49.900<a class="anchor" aria-label="anchor" href="#platjam-0049900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-49-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-49-900"></a></h3>
<ul><li>
<code><a href="../reference/save_salmon_qc_xlsx.html">save_salmon_qc_xlsx()</a></code> new argument <code>verbose=FALSE</code>, thus changing the default which included some verbose output by default.</li>
<li>
<code><a href="../reference/df_to_numcolors.html">df_to_numcolors()</a></code> new argument <code>trimRamp</code> passed to <code><a href="https://rdrr.io/pkg/jamba/man/getColorRamp.html" class="external-link">jamba::getColorRamp()</a></code> to trim edge colors from linear color gradient, thus making the colors less dramatic.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.48.900" id="platjam-0048900">platjam 0.0.48.900<a class="anchor" aria-label="anchor" href="#platjam-0048900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-48-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-48-900"></a></h3>
<ul><li>
<code><a href="../reference/design2layout.html">design2layout()</a></code> is experimental, and is intended to determine appropriate plot panel layout dimensions, given experimental groups and batch information. It should help align panels within batch, then within group, into columns, so plots can be visually compared.</li>
<li>
<code><a href="../reference/cardinality.html">cardinality()</a></code> takes two vectors and determines the data cardinality, in terms of “one-to-one”, “one-to-many”, “many-to-many”, “many-to-one”. This function may help detect when experimental factors are compatible with color assignment by group or other factors.</li>
<li>
<code><a href="../reference/df_to_numcolors.html">df_to_numcolors()</a></code> is an experimental function only used to create a color substitution vector for use with <code><a href="https://rdrr.io/pkg/jamba/man/kable_coloring.html" class="external-link">kable_coloring()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.47.900" id="platjam-0047900">platjam 0.0.47.900<a class="anchor" aria-label="anchor" href="#platjam-0047900"></a></h2>
<div class="section level3">
<h3 id="new-function-import_lipotype-0-0-47-900">new function <code>import_lipotype()</code>
<a class="anchor" aria-label="anchor" href="#new-function-import_lipotype-0-0-47-900"></a></h3>
<ul><li>Intended to import CSV formats provided by LipoType lipidomics data</li>
<li>Returns <code>SummarizedExperiment</code> object, with rowData and colData assigned as appropriate.</li>
<li>Next iteration will parse optional <code>curation_txt</code> to assign sample annotation for experimental design, factors, etc.</li>
</ul></div>
<div class="section level3">
<h3 id="new-function-design2colors-in-development-0-0-47-900">new function <code>design2colors()</code> (in development)<a class="anchor" aria-label="anchor" href="#new-function-design2colors-in-development-0-0-47-900"></a></h3>
<p>See examples <code>? design2colors()</code> for usage.</p>
<ul><li><p>Intended as an extension to <code><a href="https://rdrr.io/pkg/colorjam/man/group2colors.html" class="external-link">colorjam::group2colors()</a></code> that also utilizes subgroup (smaller subsets of groups), and group class (which includes multiple groups together).</p></li>
<li><p>If this function works well through early testing, it likely moves into <code>colorjam</code> for broader re-use.</p></li>
<li>
<p>I find myself assigning categorical colors by group, then using <code><a href="https://rdrr.io/pkg/jamba/man/color2gradient.html" class="external-link">jamba::color2gradient()</a></code> to split colors into light-to-dark gradient:</p>
<ul><li>time course, 0-hr, 2-hr, 4-hr, 8-hr (light-to-dark)</li>
<li>treatment: untreated, treated (light-to-dark)</li>
<li>for each group: batch A, batch B (light-to-dark)</li>
</ul></li>
<li>
<p>I find myself assigning these categorical colors to each SampleID, or to subgroups, or other factors, so the colors are consistent when used in tables, and <code>ComplexHeatmap</code> output.</p>
<ul><li>assign colors to all annotation columns with appropriate cardinality</li>
</ul></li>
<li>
<p>Finally, other factors need colors, they should differ from group colors</p>
<ul><li>for each individual factor, assign unique categorical colors</li>
<li>re-use colors when a factor level already has an assigned color</li>
</ul></li>
<li>
<p>Default output:</p>
<ul><li>colors assigned for every column, and every column value of the input <code>data.frame</code> argument <code>x</code>.</li>
<li>Plots a graphical table of sample annotations showing color assignments using <code><a href="https://rdrr.io/pkg/jamba/man/imageByColors.html" class="external-link">jamba::imageByColors()</a></code>.</li>
<li>
<code>list</code> of colors, named by <code>colnames(x)</code>, suitable for use with <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html" class="external-link">ComplexHeatmap::HeatmapAnnotation()</a></code>.</li>
<li>optionally a color vector, named by factor level.</li>
<li>optionally returns a <code>data.frame</code> with colors, in same order as input <code>x</code>, suitable for use directly in <code><a href="https://rdrr.io/pkg/jamba/man/imageByColors.html" class="external-link">jamba::imageByColors()</a></code>.</li>
</ul></li>
<li>
<p>Relevant arguments</p>
<ul><li><p><code>group_colnames</code>: defines categorical colors, one per group</p></li>
<li><p><code>lightness_colnames</code>: assigns light-to-dark gradient to categorical colors</p></li>
<li><p><code>class_colnames</code>: arranges groups into classes, so groups with same class are assigned similar color hues, with spacing between classes.</p></li>
<li>
<p><code>preset</code>: defines the color wheel used by <code><a href="https://rdrr.io/pkg/colorjam/man/rainbowJam.html" class="external-link">colorjam::rainbowJam()</a></code>:</p>
<ul><li>
<code>dichromat</code>: dichromat-friendly color wheel, largely removes green</li>
<li>
<code>ryb1</code>: red-yellow-blue (common painter’s color wheel for color mixing)</li>
<li>
<code>rgb</code>: red-green-blue (default R, shows how much of the wheel is green)</li>
</ul></li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.46.900" id="platjam-0046900">platjam 0.0.46.900<a class="anchor" aria-label="anchor" href="#platjam-0046900"></a></h2>
<div class="section level3">
<h3 id="new-function-0-0-46-900">new function<a class="anchor" aria-label="anchor" href="#new-function-0-0-46-900"></a></h3>
<ul><li>
<p><code><a href="../reference/import_salmon_quant.html">import_salmon_quant()</a></code></p>
<ul><li>does the job of <code><a href="https://rdrr.io/pkg/tximport/man/tximport.html" class="external-link">tximport::tximport()</a></code>, then associates sample annotations via <code>curation_txt</code> to produce <code>SummarizedExperiment</code> objects at transcript and/or gene level.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.45.900" id="platjam-0045900">platjam 0.0.45.900<a class="anchor" aria-label="anchor" href="#platjam-0045900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-45-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-45-900"></a></h3>
<ul><li>
<p><code><a href="../reference/get_salmon_meta.html">get_salmon_meta()</a></code></p>
<ul><li>now parses the fragment length distribution if the file exists. It returns mean, median, mode, and 25/75 percentiles.</li>
<li>Also the <code>names(metafile)</code> are maintaines as rownames in the output.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-45-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-45-900"></a></h3>
<ul><li>
<code><a href="../reference/parse_salmon_flenfile.html">parse_salmon_flenfile()</a></code> parses the Salmon fragment length file usually named <code>"flenDist.txt"</code>, and return summary values: mean, median, mode, 25/75 percentiles.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.44.900" id="platjam-0044900">platjam 0.0.44.900<a class="anchor" aria-label="anchor" href="#platjam-0044900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-44-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-44-900"></a></h3>
<ul><li>UCSC track hub function <code><a href="../reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> was updated because apparently <code><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue::glue()</a></code> added a new argument <code>.trim</code> to trim whitespace from every output line AND made <code>.trim=TRUE</code> the default value. Because why be consistent with past versions of <code>glue</code>. Thank you again to all the R package authors that have chosen not to change the output of the primary function in their package by default. :) Not shown: All the updates I did not have to do because other core functions keep working as normal.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.43.900" id="platjam-0043900">platjam 0.0.43.900<a class="anchor" aria-label="anchor" href="#platjam-0043900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-43-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-43-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> was updated to handle transformation of data during k-means clustering. Previously this function was not correctly defined.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.42.900" id="platjam-0042900">platjam 0.0.42.900<a class="anchor" aria-label="anchor" href="#platjam-0042900"></a></h2>
<div class="section level3">
<h3 id="updates-to-existing-functions-0-0-42-900">updates to existing functions<a class="anchor" aria-label="anchor" href="#updates-to-existing-functions-0-0-42-900"></a></h3>
<ul><li>
<code><a href="../reference/import_proteomics_PD.html">import_proteomics_PD()</a></code> gained some options to handle PD data content, such as referring Protein gene annotations into the peptide data during import. Apparently sometimes they only annotate the protein level, so this step recovers some annotations otherwise lost. Also sometimes peptide data is split across multiple rows when the peptide may appear in multiple positions in the same protein.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.41.900" id="platjam-0041900">platjam 0.0.41.900<a class="anchor" aria-label="anchor" href="#platjam-0041900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-41-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-41-900"></a></h3>
<ul><li>
<code><a href="../reference/import_proteomics_PD.html">import_proteomics_PD()</a></code> will import proteomics abundance data from Proteomics Discoverer software, focused on the Excel format in use in 2022 that includes protein abundance data, with nested sub-tables of peptide abundance data with different columns. The function will parse each type of row and return two <code>SummarizedExperiment</code> objects, one for protein data, one for peptide data. This function calls an internal helper function <code><a href="../reference/convert_PD_df_to_SE.html">convert_PD_df_to_SE()</a></code> which performs the work of parsing gene annotations, peptides, post-translational modifications, sample annotations, and optionally reads <code>"curation.txt"</code> to define specific sample group annotations for downstream analysis.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.40.900" id="platjam-0040900">platjam 0.0.40.900<a class="anchor" aria-label="anchor" href="#platjam-0040900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-40-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-40-900"></a></h3>
<ul><li>
<code><a href="../reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> was updated to allow additional parameters to be defined through <code>...</code> dot list. These values will override all values defined in the track lines.
<ul><li>Examples have been added to demonstrate minimal use of the function, showing composite track views, and multiWig overlay tracks.</li>
<li>new experimental argument <code>multiwig_concat_header=TRUE</code> allows adjusting the format of multiWig superTrack names, being tested inside the UCSC genome browser.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.39.900" id="platjam-0039900">platjam 0.0.39.900<a class="anchor" aria-label="anchor" href="#platjam-0039900"></a></h2>
<div class="section level3">
<h3 id="changes-to-nmatlist2heatmaps-0-0-39-900">changes to <code>nmatlist2heatmaps()</code>
<a class="anchor" aria-label="anchor" href="#changes-to-nmatlist2heatmaps-0-0-39-900"></a></h3>
<ul><li>
<p>The <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">ComplexHeatmap::draw()</a></code> function by default includes the argument <code>adjust_annotation_extension=TRUE</code> described here <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/a-list-of-heatmaps.html#adjust-blank-space-caused-by-annotations" class="external-link uri">https://jokergoo.github.io/ComplexHeatmap-reference/book/a-list-of-heatmaps.html#adjust-blank-space-caused-by-annotations</a></p>
<ul><li>The problem was that y-axis labels on the right side of profile plots overlapped the color legends on the right side.</li>
<li>Alternatively, profile axis labels may be positioned on the left side, which avoids having to adjust the legend on the right side.</li>
</ul></li>
<li><p>New argument <code>top_axis_side</code> with values “right”, “left”, “both”, “all”, “none”</p></li>
<li><p>New argument <code>k_heatmap</code> which defines which heatmap or heatmaps are used for k-means clustering. Key new feature, k-means clustering can now involve multiple heatmaps, which may be useful to find patterns across multiple signals. This feature is being tested currently.</p></li>
<li><p><code><a href="../reference/save_salmon_qc_xlsx.html">save_salmon_qc_xlsx()</a></code> was updated to include jamba:: function prefixes as needed.</p></li>
</ul></div>
<div class="section level3">
<h3 id="functions-removed-0-0-39-900">functions removed<a class="anchor" aria-label="anchor" href="#functions-removed-0-0-39-900"></a></h3>
<ul><li>
<code><a href="https://rdrr.io/pkg/jamba/man/set_xlsx_colwidths.html" class="external-link">set_xlsx_colwidths()</a></code> and <code><a href="https://rdrr.io/pkg/jamba/man/set_xlsx_rowheights.html" class="external-link">set_xlsx_rowheights()</a></code> were removed, since they have been moved into package <code>"jamba"</code>.</li>
<li>
<code>twostep_gradient()</code>, <code>showDichromat()</code>, and <code>make_jam_divergent()</code> were moved into package <code>"colorjam"</code>.</li>
<li>
<code>jam_divergent</code> and <code>jam_linear</code> were moved into package <code>"colorjam"</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.38.900" id="platjam-0038900">platjam 0.0.38.900<a class="anchor" aria-label="anchor" href="#platjam-0038900"></a></h2>
<div class="section level3">
<h3 id="changes-to-nmatlist2heatmaps-0-0-38-900">changes to <code>nmatlist2heatmaps()</code>
<a class="anchor" aria-label="anchor" href="#changes-to-nmatlist2heatmaps-0-0-38-900"></a></h3>
<p>A series of changes was made to <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>:</p>
<ul><li>
<p>New argument <code>top_annotation</code> to customize the annotation appearing above each heatmap. It takes several forms:</p>
<ol style="list-style-type: decimal"><li>
<code>TRUE</code> uses the default <code>ComplexHeatmap::HeatmapAnnotation(EnrichedHeatmap::anno_enriched())</code> with some platjam-specific customizations, mainly to match the profile line colors to the <code>partition</code> and <code>k_colors</code> arguments for consistency.</li>
<li>
<code>FALSE</code> hides the annotation</li>
<li>
<code>HeatmapAnnotation</code> uses the function as provided</li>
<li>
<code>list</code> of any combination of the above, recycled to <code>length(nmatlist)</code>. The <code>list</code> input can be used for specific plots above each matrix, which can also specify the data used for those plots. Note option 3 above uses the closed form <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html" class="external-link">EnrichedHeatmap::anno_enriched()</a></code> in order to be applied to the appropriate data matrix for each heatmap.</li>
</ol></li>
<li><p>New argument <code>top_anno_height</code> which is a <code><a href="https://rdrr.io/r/grid/unit.html" class="external-link">grid::unit</a></code> object specifying the height of the <code>top_annotation</code> - only used when <code>top_annotation</code> is not supplied in custom form.</p></li>
<li>
<p>Help text for argument <code>signal_ceiling</code> includes reference to the child function <code><a href="../reference/get_nmat_ceiling.html">get_nmat_ceiling()</a></code> which controls the logic:</p>
<ul><li>
<code>signal_ceiling</code> above <code>1</code> applies the strict numeric value</li>
<li>
<code>signal_ceiling</code> between <code>0</code> and <code>1</code> applies a <code>quantile(x, probs=signal_ceiling)</code>
</li>
<li>
<code>signal_ceiling=NULL</code> uses the maximum absolute value</li>
</ul></li>
<li><p>Argument <code>row_order</code> is intended to work alongside <code>byCols</code>, so the sort of <code>anno_df</code> will also inlude sort for row order within each subset of <code>anno_df</code> as applicable.</p></li>
<li><p>New arguments <code>raster_quality</code> and <code>raster_by_magick</code> passed to <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></li>
<li><p>New arguments <code>upstream_length</code>, <code>downstream_length</code> used for optional zoom of the coordinate range being displayed. These arguments are passed to <code><a href="../reference/zoom_nmatlist.html">zoom_nmatlist()</a></code> if either is not <code>NULL</code>.</p></li>
<li>
<p>New behavior when <code>panel_groups</code> are supplied:</p>
<ul><li>
<code>heatmap_annotation_param</code> will automatically hide duplicate color legends, and will use the <code>panel_groups</code> value for the color legend title.</li>
<li>
<code>top_annotation</code> default axis will be hidden except for the last in each set of adjacent duplicated panel_groups. For example, <code>panel_groups=c("A", "A", "B", "B", "A", "A")</code> will display y-axis for entries <code>c(2, 4, 6)</code>.</li>
<li>In future, heatmap panels may have slightly wider gap between different panel_groups values.</li>
</ul></li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-38-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-38-900"></a></h3>
<ul><li>
<code><a href="../reference/zoom_nmatlist.html">zoom_nmatlist()</a></code> and <code><a href="../reference/zoom_nmat.html">zoom_nmat()</a></code> are intended to help zoom into a particular range of coverage matrix data, where the stored coverage is wider than desired for the heatmap figure. They filter the columns in each <code>normalizedMatrix</code> and adjust the attributes stored in:
<ul><li>
<code>attr(nmat, "upstream_index")</code> - the column index positions upstream the target region</li>
<li>
<code>attr(nmat, "downstream_index")</code> - the column index positions downstream the target region</li>
<li>
<code>attr(nmat, "target_index")</code> - the column index positions representing the target region</li>
<li>
<code>attr(nmat, "extend")</code> - the genomic distance upstream and downstream the target region</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.37.900" id="platjam-0037900">platjam 0.0.37.900<a class="anchor" aria-label="anchor" href="#platjam-0037900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-37-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-37-900"></a></h3>
<ul><li>
<code><a href="../reference/get_track_defaults.html">get_track_defaults()</a></code> was updated to adjust the default bigBed track settings, so the default display is <code>"pack"</code> but can be adjusted as needed.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.36.900" id="platjam-0036900">platjam 0.0.36.900<a class="anchor" aria-label="anchor" href="#platjam-0036900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-36-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-36-900"></a></h3>
<ul><li>
<code><a href="../reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> handles bigBed tracks with specific composite track templates and relevant default values.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.35.900" id="platjam-0035900">platjam 0.0.35.900<a class="anchor" aria-label="anchor" href="#platjam-0035900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-35-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-35-900"></a></h3>
<ul><li>
<code><a href="../reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> was updated to improve the method of naming supergroup and group tracks, so the supergroup name will persist for every group it should contain.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>trim_legend_title</code> will trim heatmap legend title to remove any text following a newline. For example, the heatmap label may contain multiple lines of text which are useful to the heatmap, but less useful to the color legend. In that case, only the first line of text is used as the heatmap legend title.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-35-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-35-900"></a></h3>
<ul><li>
<code><a href="../reference/nmathm_row_order.html">nmathm_row_order()</a></code> returns the rowname order of the heatmap produced by <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>. When rows are partitioned, the output is a <code>list</code> of rowname vectors.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.34.900" id="platjam-0034900">platjam 0.0.34.900<a class="anchor" aria-label="anchor" href="#platjam-0034900"></a></h2>
<div class="section level3">
<h3 id="new-additions-0-0-34-900">new additions<a class="anchor" aria-label="anchor" href="#new-additions-0-0-34-900"></a></h3>
<ul><li>
<code>jam_linear</code> and <code>jam_divergent</code> are new color gradients intended to be visibly distinct, and color-blind-friendly. The divergent scales in particular are designed so that the two directions are visibly distinct even under three different color blindness simulations from <code><a href="https://rdrr.io/pkg/dichromat/man/dichromat.html" class="external-link">dichromat::dichromat()</a></code>.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-34-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-34-900"></a></h3>
<ul><li><p><code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> changes that forced heatmap legend direction to be <code>"vertical" were reverted, since Dr. Gu updated</code>ComplexHeatmap<code>to fix the issue with</code>grid` on R version 3.6.*.</p></li>
<li>
<p><code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>:</p>
<ul><li>new argument <code>pos_line</code> to control display of lines around the target region</li>
<li>return values include <code>HM_drawn</code> which is the object returned after <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">ComplexHeatmap::draw()</a></code>, and which contains all the column and row orders for each heatmap.</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.33.900" id="platjam-0033900">platjam 0.0.33.900<a class="anchor" aria-label="anchor" href="#platjam-0033900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-33-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-33-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> encountered an error when displaying heatmap legend with <code>direction="horizontal"</code> which appears to be specific to R version R-3.6.1, and is not present in R-4.0.0 and above. Specifically I believe this error is from <code>grid</code> package version <code>3.6.1</code> and how the function <code>unit.arithmetic()</code> is called by ComplexHeatmap. Nonetheless, the workaround is to force <code>direction="vertical"</code> when <code>grid</code> package version is below <code>4.0.0</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.32.900" id="platjam-0032900">platjam 0.0.32.900<a class="anchor" aria-label="anchor" href="#platjam-0032900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-32-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-32-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>show_heatmap_legend</code> is recycled to the <code>length(nmatlist)</code> to allow hiding legends for certain heatmaps. This update is intended when heatmaps are grouped by <code>panel_groups</code> which allows multiple heatmaps to have the same shared numeric color mapping.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.31.900" id="platjam-0031900">platjam 0.0.31.900<a class="anchor" aria-label="anchor" href="#platjam-0031900"></a></h2>
<div class="section level3">
<h3 id="updates-0-0-31-900">updates<a class="anchor" aria-label="anchor" href="#updates-0-0-31-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> updated so <code>ylims</code> will take priority over <code>panel_groups=TRUE</code> which defines its own ylims ranges.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.28.900" id="platjam-0028900">platjam 0.0.28.900<a class="anchor" aria-label="anchor" href="#platjam-0028900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-28-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-28-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> bug with no <code>k_clusters</code>, no <code>partition</code>, and no <code>k_colors</code> would fail during profile plot because it had NULL color.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.28.900" id="platjam-0028900-1">platjam 0.0.28.900<a class="anchor" aria-label="anchor" href="#platjam-0028900-1"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-28-900-1">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-28-900-1"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> does more work to validate <code>rows</code> with respect to rownames of <code>nmatlist</code>, <code>rownames(anno_df)</code>, and <code>names(partition)</code> to ensure everything is properly synchronized. The smallest common set of <code>rows</code> will be used.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-28-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-28-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> now allows <code>partition</code> and <code>k_clusters</code> to be combined, so rows can be both partitioned, and kmeans clustered. Labels are created <code>"partition - cluster_number"</code> and are ordered by partition then kmeans cluster number. So if <code>partition</code> is a factor with ordered levels, that level will be preserved in the output.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>color_sub</code> which allows defined categorical colors by sending a named vector. Any column in <code>anno_df</code> all of whose values are in <code>names(color_sub)</code> will be colored accordingly.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> will use <code>k_subset</code> to re-order partitions, in addition to plotting only the subset of row partitions. The code was change to handle the sequence of operations: kmeans, partitioning, k_subset, then anno_df, marked rows.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> changed the default <code>ht_gap</code> down to 3 mm.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.27.900" id="platjam-0027900">platjam 0.0.27.900<a class="anchor" aria-label="anchor" href="#platjam-0027900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-27-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-27-900"></a></h3>
<ul><li>Numerous small bug fixes in <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>, mostly in handling annotation categorical color legends.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.26.900" id="platjam-0026900">platjam 0.0.26.900<a class="anchor" aria-label="anchor" href="#platjam-0026900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-26-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-26-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new default <code>transform="none"</code>, previously was <code><a href="https://rdrr.io/pkg/jamba/man/log2signed.html" class="external-link">jamba::log2signed()</a></code> which enforced log2 transformation. Frankly not sure which is best, might depend upon the range of data, which depends upon the type of coverage signal being used.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> now treats annotation columns in <code>anno_df</code> as bi-directional when the max value is &lt;= 50, which helps in cases that have values <code>c(-1, 0, 1)</code> and sometimes only <code>c(0, 1)</code>. The end result is more numeric columns will have the same color gradient (blue-white-red).</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> displays legend with distinct color steps for <code>anno_df</code> columns with 10 or fewer distinct values, which is mostly helpful with a small number of integer values.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> uses <code>nrow</code> for discrete annotation color layout instead of <code>ncol</code> – jokergoo::ComplexHeatmap kindly updated the package to handle both, but nrow is more definitive in this use case.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.25.900" id="platjam-0025900">platjam 0.0.25.900<a class="anchor" aria-label="anchor" href="#platjam-0025900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-25-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-25-900"></a></h3>
<ul><li>
<code><a href="../reference/applyXlsxConditionalFormatByColumn.html">applyXlsxConditionalFormatByColumn()</a></code> fixed bug in verbose output that referred to nonexistent object. Also removed <code>verbose=TRUE</code> from calling function <code><a href="../reference/save_salmon_qc_xlsx.html">save_salmon_qc_xlsx()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.24.900" id="platjam-0024900">platjam 0.0.24.900<a class="anchor" aria-label="anchor" href="#platjam-0024900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-24-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-24-900"></a></h3>
<ul><li>
<code><a href="../reference/save_salmon_qc_xlsx.html">save_salmon_qc_xlsx()</a></code> saves output from <code><a href="../reference/get_salmon_meta.html">get_salmon_meta()</a></code> in a convenient Excel <code>.xlsx</code> format, with four worksheets showing different aspects of data quality. It returns a <code>list</code> of <code>data.frame</code> summaries invisibly, for further review.</li>
<li>
<code><a href="../reference/applyXlsxConditionalFormatByColumn.html">applyXlsxConditionalFormatByColumn()</a></code> applies conditional formatting using values in each column, rather than using a fixed range for all columns. This function will become part of <code><a href="https://rdrr.io/pkg/jamba/man/writeOpenxlsx.html" class="external-link">jamba::writeOpenxlsx()</a></code> and <code><a href="https://rdrr.io/pkg/jamba/man/applyXlsxConditionalFormat.html" class="external-link">jamba::applyXlsxConditionalFormat()</a></code> in the near future.</li>
<li>
<code><a href="https://rdrr.io/pkg/jamba/man/set_xlsx_colwidths.html" class="external-link">set_xlsx_colwidths()</a></code> and <code><a href="https://rdrr.io/pkg/jamba/man/set_xlsx_rowheights.html" class="external-link">set_xlsx_rowheights()</a></code> define fixed column width and row height, respectively, for <code>.xlsx</code> files.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.23.900" id="platjam-0023900">platjam 0.0.23.900<a class="anchor" aria-label="anchor" href="#platjam-0023900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-23-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-23-900"></a></h3>
<ul><li>
<code><a href="../reference/color_complement.html">color_complement()</a></code> takes input color, rotates the color hue 180 degrees to make a complement color. It is unique in that it uses warped hue (see <code><a href="https://rdrr.io/pkg/colorjam/man/h2hw.html" class="external-link">colorjam::h2hw()</a></code>) which therefore uses red-yellow-blue color wheel instead of red-green-blue, which is much better for color operations like these. This function will almost certainly be moved into the <code>colorjam</code> package. The examples show how to make a divergent color gradient from a single color.</li>
</ul></div>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-23-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-23-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>nmat_names</code> used to supply custom names for each heatmap, otherwise the attribute <code>"signal_name"</code> values are used.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> fixed error caused by <code>use_raster=TRUE</code> and any operation that expects to create a grid graphical object, such as <code><a href="https://rdrr.io/pkg/multienrichjam/man/grid_with_title.html" class="external-link">multienrichjam::grid_with_title()</a></code>, or multi-row output specified with <code>hm_nrow=2</code>.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new arguments <code>heatmap_legend_param</code> and <code>annotation_legend_param</code> allow customizing the legends in detail, passed to <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>. However, default settings show color gradients horizontally, which may save some screen space.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new arguments <code>title</code> and <code>caption</code> invoke a prototype method that calls <code><a href="https://rdrr.io/pkg/multienrichjam/man/grid_with_title.html" class="external-link">multienrichjam::grid_with_title()</a></code> to position an overall title and caption above and below the heatmaps, respectively. When supplied, these sections are displayed and the heatmaps are resized accordingly.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> now properly handles <code>k_subset</code> to plot a subset of clusters provided by <code>partition</code> or after <code>k_clusters</code> is used to create clusters. Mostly useful for follow-up plots to drill down into one or more specific clusters.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> now properly handles <code>row_order</code> for edge cases like <code>k_subset</code>, or when <code>row_order=NULL</code>; also <code>row_order=TRUE</code> will enable the default <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/enriched_score.html" class="external-link">EnrichedHeatmap::enriched_score()</a></code> for the <code>main_heatmap</code> using only the rows being displayed.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>border</code> whether to draw a border around the heatmap. It can be supplied as a vector, to control the border around each heatmap if needed.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.22.900" id="platjam-0022900">platjam 0.0.22.900<a class="anchor" aria-label="anchor" href="#platjam-0022900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-22-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-22-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> accepts input <code>nmat_colors</code> with color functions, for example sending output from <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">circlize::colorRamp2()</a></code>.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> when given a single color per matrix, when data is determined to have positive and negative values, it generates a divergent color ramp using the supplied color, and complementary color. To define a specific set of colors, supply a three color vector, the middle color will be fixed at zero.</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-0-22-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-22-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> no longer requires <code>ylims</code> and uses <code>ylim=NULL</code> properly.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.21.900" id="platjam-0021900">platjam 0.0.21.900<a class="anchor" aria-label="anchor" href="#platjam-0021900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-21-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-21-900"></a></h3>
<ul><li>
<code><a href="../reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> was updated to recognize bigBed/bb files can not become overlay tracks.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.20.900" id="platjam-0020900">platjam 0.0.20.900<a class="anchor" aria-label="anchor" href="#platjam-0020900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-20-900">changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-20-900"></a></h3>
<ul><li>
<code><a href="../reference/get_track_defaults.html">get_track_defaults()</a></code> was modified to include <code>shortLabel,longLabel</code> in the template for multiWig tracks. Apparently a label is displayed and it uses the internal hub number as a prefix by default.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.19.900" id="platjam-0019900">platjam 0.0.19.900<a class="anchor" aria-label="anchor" href="#platjam-0019900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-19-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-19-900"></a></h3>
<ul><li>
<code><a href="../reference/get_salmon_meta.html">get_salmon_meta()</a></code> is intended to load Salmon transcript quantification metadata from several output JSON files, returning results in a <code>data.frame</code> with one row per Salmon run. It combines various measured statistics from <code>"meta_info.json"</code>, <code>"cmd_info.json"</code>, and <code>"lib_format_counts.json"</code> so that multiple Salmon runs can be directly compared in a table format.</li>
<li>
<code><a href="../reference/get_salmon_root.html">get_salmon_root()</a></code> is a wrapper to <code><a href="https://rprojroot.r-lib.org/reference/find_root.html" class="external-link">rprojroot::find_root()</a></code>, intended to find the root directory of a Salmon run given any Salmon output file. It searches for the root directory that contains the file <code>"cmd_info.json"</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.18.900" id="platjam-0018900">platjam 0.0.18.900<a class="anchor" aria-label="anchor" href="#platjam-0018900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-18-900">Changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-18-900"></a></h3>
<ul><li>
<code><a href="../reference/get_numeric_transform.html">get_numeric_transform()</a></code> was updated to clean up the help docs, to accept <code>"linear"</code> as a valid name (which does not change the input data), and to parse the input <code>transform</code> values in order to assign reasonable names where possible.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.17.900" id="platjam-0017900">platjam 0.0.17.900<a class="anchor" aria-label="anchor" href="#platjam-0017900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-17-900">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-17-900"></a></h3>
<ul><li>Fixed various bugs in <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>. No doubt there will be more, covering the host of assumptions made for the panel_groups, signal_ceiling, and ylims arguments.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.16.900" id="platjam-0016900">platjam 0.0.16.900<a class="anchor" aria-label="anchor" href="#platjam-0016900"></a></h2>
<div class="section level3">
<h3 id="changes-to-nmatlist2heatmaps-0-0-16-900">Changes to <code>nmatlist2heatmaps()</code>
<a class="anchor" aria-label="anchor" href="#changes-to-nmatlist2heatmaps-0-0-16-900"></a></h3>
<ul><li>Cool new feature: <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument <code>panel_group</code>, when it is defined, it calculates a shared <code>ylims</code> for each group of panels, and a shared <code>signal_ceiling</code> for each group of panels. The signal ceiling uses the supplied <code>signal_ceiling</code> for each panel (if supplied) then takes the maximum value for each group of panels. So if <code>signal_ceiling=0.8</code> it is treated as quantile 0.8, the 0.8 quantile is calculate for all non-zero values for each panel, then each group of panels uses the max for that group.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> the <code>ht_gap</code> gap between heatmaps is not applied between annotation heatmaps, instead <code>"1mm"</code> gap is used between annotation heatmaps. The purpose of <code>ht_gap</code> is to have enough gap between heatmaps so the y-axis profile plot labels can be displayed, but there are no y-axis labels for annotations.</li>
<li>The default <code>lens</code> for <code>anno_df</code> numeric columns is now <code>lens=1</code>, reverting previous change <code>lens=2</code> which was causing gradients to be darker than expected.</li>
<li>The verbose output is now more concise, and TRUE by default. For even more detail use <code>verbose=2</code>.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-16-900">New functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-16-900"></a></h3>
<ul><li>
<code><a href="../reference/get_nmat_ceiling.html">get_nmat_ceiling()</a></code> is a simple wrapper function called by <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> and not intended to be called directly. It takes a numeric matrix, and ceiling, and returns what I determines to be the appropriate ceiling to use.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.15.900" id="platjam-0015900">platjam 0.0.15.900<a class="anchor" aria-label="anchor" href="#platjam-0015900"></a></h2>
<div class="section level3">
<h3 id="changes-to-existing-functions-0-0-15-900">Changes to existing functions<a class="anchor" aria-label="anchor" href="#changes-to-existing-functions-0-0-15-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> modified <code>signal_name</code> to clean up leading/trailing whitespace and return <code>\n</code>, and repeated return <code>\n</code> characters.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>ht_gap</code> used to define the gap between heatmap panels. This argument is intended to help reduce the overlap of the top y-axis profile labels in adjacent profile panels.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>axis_name</code>, as either a vector, or a list of vectors, used for x-axis labels across each panel.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>column_title_gp</code>, as either a <code><a href="https://rdrr.io/r/grid/gpar.html" class="external-link">grid::gpar</a></code> object, or list of gpar objects.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>profile_value</code>, passed to <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html" class="external-link">EnrichedHeatmap::anno_enriched()</a></code> to define the type of numeric summary to use for the profile plots.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>panel_groups</code>, character vector of groups used to apply consistent y-axis ylims values per group. By default if <code>nmat_colors</code> is not supplied, then <code>colorjam::group2colors(panel_groups)</code> is used to define group colors.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.14.900" id="platjam-0014900">platjam 0.0.14.900<a class="anchor" aria-label="anchor" href="#platjam-0014900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-14-900">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-14-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> fixed bug in handling argument <code>signal_ceiling</code>, which when the value is between 0 and 1 is treated as a quantile. The method now removes zero before the <code><a href="https://rdrr.io/r/stats/quantile.html" class="external-link">quantile()</a></code> calculation.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> fixed bug that wrongly expanded argument <code>axis_names_gp</code> as a list without realizing class <code>"gpar"</code> is also a list. Now <code>"gpar"</code> is wrapped in <code><a href="https://rdrr.io/r/base/list.html" class="external-link">list()</a></code> before expanding to the number of coverage matrices.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> fixed bug where argument <code>axis_name_rot</code> was not defined in the function itself.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.13.900" id="platjam-0013900">platjam 0.0.13.900<a class="anchor" aria-label="anchor" href="#platjam-0013900"></a></h2>
<div class="section level3">
<h3 id="enhancements-0-0-13-900">Enhancements<a class="anchor" aria-label="anchor" href="#enhancements-0-0-13-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> altered the logic used to define color ranges in numeric columns of <code>anno_df</code>, opting to use quantile ranges <code>c(0.005, 0.995)</code> to help trim extreme values. Also changed <code>"purple"</code> to <code>"Purples"</code> to use the <code>RColorBrewer</code> purple color gradient.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.12.900" id="platjam-0012900">platjam 0.0.12.900<a class="anchor" aria-label="anchor" href="#platjam-0012900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-12-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-12-900"></a></h3>
<ul><li>
<code><a href="../reference/get_numeric_transform.html">get_numeric_transform()</a></code> bug fixed wrong argument name used inside the function. Should also resolve issues with <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.11.900" id="platjam-0011900">platjam 0.0.11.900<a class="anchor" aria-label="anchor" href="#platjam-0011900"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-0-11-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-11-900"></a></h3>
<ul><li>
<code><a href="../reference/import_nanostring_rcc.html">import_nanostring_rcc()</a></code> reads Nanostring RCC files, and produces a <code>SummarizedExperiment</code> object suitable for use in omics-style analyses.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.10.900" id="platjam-0010900">platjam 0.0.10.900<a class="anchor" aria-label="anchor" href="#platjam-0010900"></a></h2>
<div class="section level3">
<h3 id="enhancements-0-0-10-900">enhancements<a class="anchor" aria-label="anchor" href="#enhancements-0-0-10-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>signal_ceiling</code> defines a numerical ceiling for each input matrix, used to apply the color ramp to a defined numeric range. When <code>signal_ceiling</code> is between 0 and 1, it is assumed to be a quantile, so the appropriate quantile value is used as the ceiling. These values can be seen when <code>verbose=TRUE</code>.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument <code>transform</code> now accepts text string for transformation functions, such as <code>"log2signed"</code>, <code>"sqrt"</code>, <code>"cubert"</code>, etc. Any other string will try to match a function name in the R search path.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-10-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-10-900"></a></h3>
<ul><li>
<code><a href="../reference/get_numeric_transform.html">get_numeric_transform()</a></code> is used to apply name-based transformations, called from <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>.</li>
</ul></div>
<div class="section level3">
<h3 id="ucsc-genome-browser-functions-0-0-10-900">UCSC genome browser functions<a class="anchor" aria-label="anchor" href="#ucsc-genome-browser-functions-0-0-10-900"></a></h3>
<p>Some new functions were added to help create UCSC genome browser track hubs, which are fairly tedious to build manually.</p>
<ul><li>
<code><a href="../reference/get_track_defaults.html">get_track_defaults()</a></code> returns default values for tracks, including some track snippet templates for composite and overlay track types.</li>
<li>
<code><a href="../reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> is a parser for a specific track type, based upon prominent scientist Dr. Gokey, who defines a straightforward format for a series of grouped tracks.</li>
<li>
<code><a href="../reference/make_ucsc_trackname.html">make_ucsc_trackname()</a></code> creates a valid UCSC track name, removing invalid characters.</li>
<li>
<code><a href="../reference/assign_track_defaults.html">assign_track_defaults()</a></code> assigns track values to an environment which may already contain default values. New values override previous values, after which the environment can be used with the <code>glue</code> package to update a track snippet template for each track in a set.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.9.900" id="platjam-009900">platjam 0.0.9.900<a class="anchor" aria-label="anchor" href="#platjam-009900"></a></h2>
<div class="section level3">
<h3 id="enhancements-0-0-9-900">enhancements<a class="anchor" aria-label="anchor" href="#enhancements-0-0-9-900"></a></h3>
<ul><li>
<code><a href="../reference/coverage_matrix2nmat.html">coverage_matrix2nmat()</a></code> now accepts a vector of files as input, and will process them and return a list of normalizedMatrix objects, ready to go directly into <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.8.900" id="platjam-008900">platjam 0.0.8.900<a class="anchor" aria-label="anchor" href="#platjam-008900"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-8-900">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-8-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> fixed bug in placement of argument show_error, now properly placed inside <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html" class="external-link">EnrichedHeatmap::anno_enriched()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.7.900" id="platjam-007900">platjam 0.0.7.900<a class="anchor" aria-label="anchor" href="#platjam-007900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-7-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-7-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new argument <code>show_error</code> is passed to <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html" class="external-link">EnrichedHeatmap::anno_enriched()</a></code> which enables shaded error bars using standard error around the profile at the top of each heatmap.</li>
</ul></div>
<div class="section level3">
<h3 id="new-functions-0-0-7-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-7-900"></a></h3>
<ul><li>
<code><a href="../reference/frequency_matrix2nmat.html">frequency_matrix2nmat()</a></code> converts a frequency matrix whose colnames represent numeric frequency values, and converts to class <code>"normalizedMatrix"</code> for use in EnrichedHeatmap. It optionally scales each row using <code><a href="../reference/rowNormScale.html">rowNormScale()</a></code>.</li>
<li>
<code><a href="../reference/rowNormScale.html">rowNormScale()</a></code> applies <code><a href="https://rdrr.io/pkg/jamba/man/normScale.html" class="external-link">jamba::normScale()</a></code> to each row of a matrix.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.6.900" id="platjam-006900">platjam 0.0.6.900<a class="anchor" aria-label="anchor" href="#platjam-006900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-6-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-6-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument <code>k_subset</code> now works when <code>partition</code> is supplied and k-means clustering is not performed inside this function. Therefore, one can supply their own clusters, and choose a subset for subsequent drilldown. The colors are assigned by total cluster, so the subset should retain consistent colors per cluster.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument <code>k_method</code> allows defining different k-means clustering methods, including <code>c("pearson","correlation")</code> if the <code>"amap"</code> package is installed.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.5.900" id="platjam-005900">platjam 0.0.5.900<a class="anchor" aria-label="anchor" href="#platjam-005900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-5-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-5-900"></a></h3>
<ul><li>Version bump while debugging logistics with <code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument <code>transform</code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.4.900" id="platjam-004900">platjam 0.0.4.900<a class="anchor" aria-label="anchor" href="#platjam-004900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-4-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-4-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument <code>transform</code> is now applied to each heatmap, with <code>transform</code> values recycled to the <code>length(nmatlist)</code>, to allow control over mathematical transformation of each numeric matric.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.3.900" id="platjam-003900">platjam 0.0.3.900<a class="anchor" aria-label="anchor" href="#platjam-003900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-3-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-3-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument anno_df is a data.frame with annotations that can be used to order rows in the data.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> updated its method to handle proper order of user-supplied partition information.</li>
<li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> new arguments to handle row labeling, still testing the combination of partitions and row labels which appear misaligned.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.2.900" id="platjam-002900">platjam 0.0.2.900<a class="anchor" aria-label="anchor" href="#platjam-002900"></a></h2>
<div class="section level3">
<h3 id="changes-0-0-2-900">changes<a class="anchor" aria-label="anchor" href="#changes-0-0-2-900"></a></h3>
<ul><li>The package itself allows installation on R-3.0.0 and higher, resolving an issue with installing on R-3.5.0 when the package said it required R-3.5.1 (and did not).</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.2.900" id="platjam-002900-1">platjam 0.0.2.900<a class="anchor" aria-label="anchor" href="#platjam-002900-1"></a></h2>
<div class="section level3">
<h3 id="new-function-0-0-2-900">new function<a class="anchor" aria-label="anchor" href="#new-function-0-0-2-900"></a></h3>
<ul><li>
<code><a href="../reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> creates multiple coverage heatmaps as the result of importing data with <code><a href="../reference/coverage_matrix2nmat.html">coverage_matrix2nmat()</a></code>.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.1.900" id="platjam-001900">platjam 0.0.1.900<a class="anchor" aria-label="anchor" href="#platjam-001900"></a></h2>
<p>Initial release.</p>
<div class="section level3">
<h3 id="new-functions-0-0-1-900">new functions<a class="anchor" aria-label="anchor" href="#new-functions-0-0-1-900"></a></h3>
<ul><li>
<code><a href="../reference/coverage_matrix2nmat.html">coverage_matrix2nmat()</a></code> imports genome coverage matrix file and converts to class <code>normalizedMatrix</code> for use in the amazing <code>"EnrichedHeatmap"</code> package.</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

