<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>TODO 21sep2023 • platjam</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css"><script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="pkgdown.js"></script><meta property="og:title" content="TODO 21sep2023"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-title-body">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">platjam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.73.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="http://github.com/jmw86069/platjam/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>TODO 21sep2023</h1>
    </div>


<div id="todo-21sep2023" class="section level1">

<ul><li>DONE. Update <code><a href="reference/design2colors.html">design2colors()</a></code> consistent with colorjam 0.0.26.900 changes.</li>
</ul></div>
<div class="section level1">
<h1 id="todo-31jul2023">TODO 31jul2023<a class="anchor" aria-label="anchor" href="#todo-31jul2023"></a></h1>
<ul><li>
<p><code><a href="reference/import_salmon_quant.html">import_salmon_quant()</a></code></p>
<ul><li>be more tolerant of being given paths to <code>quant.sf</code> files which may be renamed to different filenames.</li>
<li>be more tolerant when GTF and tx2gene are not supplied, the import can still import transcript-level data without adding gene-level annotation and summaries.</li>
</ul></li>
<li>
<p><code><a href="reference/design2colors.html">design2colors()</a></code></p>
<ul><li>DONE. accept <code>DataFrame</code> input, as from <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">SummarizedExperiment::colData()</a></code>.</li>
<li>DONE. use workaround for class/group cardinality different than 1-to-N</li>
<li>DONE. when lightness is defined and group/lightness cardinality is not 1-to-N, assign “grey45” to lightness_colnames in color_sub.</li>
<li>DONE. handle using <code>"-"</code> prefix for colnames to reverse the sort order</li>
<li>DONE. fix use case with no values assigned from <code>group_colnames</code>
</li>
<li>fix use case with only one input colname, no rownames are printed</li>
<li>consider sorting column assignment based upon cardinality with <code>group_colnames</code>, then number of unique entries, in order to impose some reproducibility even when input column order is changed.</li>
<li>accept color assignment for <code>class_colnames</code> values, which would then impose constraints on group color hues within reasonable range of those colors. It would effective split the color hue into a slightly wider range of hues, centered on the class color hue.</li>
</ul></li>
</ul></div>
<div class="section level1">
<h1 id="todo-19jul2023">TODO 19jul2023<a class="anchor" aria-label="anchor" href="#todo-19jul2023"></a></h1>
<ul><li>
<p>DONE: add importer for metabolomics data</p>
<ul><li>
<code><a href="reference/import_metabolomics_niehs.html">import_metabolomics_niehs()</a></code> - import LC-MS or LC-MS/MS data processed by the NIEHS Metabolomics Core Facility, with defined file formats.</li>
</ul></li>
</ul></div>
<div class="section level1">
<h1 id="todo-18apr2023">TODO 18apr2023<a class="anchor" aria-label="anchor" href="#todo-18apr2023"></a></h1>
<ul><li>
<p><code><a href="reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code></p>
<ul><li>
<p>add method to display group labels above heatmaps:</p>
<ul><li>
<code>signal</code> - (H3K27ac, H3K4me3, etc) across contiguous <code>panel_groups</code>
</li>
<li>
<code>type</code> - (signal, difference) across contiguous signal types within <code>panel_groups</code>
</li>
<li>
<code>label</code> - trimmed label which no longer requires the <code>signal</code> and <code>type</code> encoded into the name.</li>
<li>use <code>heatmap_column_group_labels()</code> style</li>
</ul></li>
<li><p>add method to adjust padding (whitespace) around the overall heatmap layout; between heatmaps and color legend; between annotation stripes.</p></li>
<li><p>make the color assignment more similar to <code><a href="https://rdrr.io/pkg/jamses/man/heatmap_se.html" class="external-link">jamses::heatmap_se()</a></code>, in the form of color <code>color_list</code> named by column.</p></li>
</ul></li>
</ul></div>
<div class="section level1">
<h1 id="todo-22feb2023">TODO 22feb2023<a class="anchor" aria-label="anchor" href="#todo-22feb2023"></a></h1>
<ul><li>
<p><code><a href="reference/design2colors.html">design2colors()</a></code></p>
<ul><li><p>Error is thrown when there are empty values in <code>group_colnames</code> fields.</p></li>
<li><p>DONE: Allow using color ramp name instead of color for column assignment.</p></li>
<li>
<p>DONE: Allow defining a <code>color_max</code>, maximum value for <code>numeric</code> columns that will use a color gradient.</p>
<ul><li>This option will help whenever a few columns should share the same threshold. It is not auto-detected, but can still be useful. The analyst will need to provide the <code>color_max</code> value upfront, perhaps in the form of a named vector similar to <code>color_sub</code>, named by <code>colnames(x)</code>.</li>
</ul></li>
<li><p>Allow <code>"rownames"</code> as valid input to <code>group_colnames</code>, essentially assigning one value per row.</p></li>
<li><p>Make the function work when <code>group_colnames</code> is <code>NULL</code>. Assign colors to rownames?</p></li>
<li><p>consider removing <code>class_group_color</code> and <code>class_group_lightness_color</code> from output, or renaming each to use the combination of actual colnames assembled to form those colors.</p></li>
</ul></li>
</ul></div>
<div class="section level1">
<h1 id="todo-10jan2023">TODO 10jan2023<a class="anchor" aria-label="anchor" href="#todo-10jan2023"></a></h1>
<ul><li>
<code><a href="reference/get_salmon_meta.html">get_salmon_meta()</a></code> - please fix the file ordering as described below.</li>
</ul></div>
<div class="section level1">
<h1 id="todo-20sep2022">TODO 20sep2022<a class="anchor" aria-label="anchor" href="#todo-20sep2022"></a></h1>
<ul><li>
<code><a href="reference/get_salmon_meta.html">get_salmon_meta()</a></code> should return data in the same order the filenames were provided, in fact it’s weird that it would not already do that.</li>
</ul></div>
<div class="section level1">
<h1 id="todo-13sep2022">TODO 13sep2022<a class="anchor" aria-label="anchor" href="#todo-13sep2022"></a></h1>
<ul><li>(COMPLETE) Add import function for Mascot proteomics type data format.</li>
</ul></div>
<div class="section level1">
<h1 id="todo-13may2022">TODO 13may2022<a class="anchor" aria-label="anchor" href="#todo-13may2022"></a></h1>
<ul><li>
<p><code><a href="reference/design2colors.html">design2colors()</a></code></p>
<ul><li>make it work even without <code>group_colnames</code>, which is effectively just colorizing a <code>data.frame</code> by each column.</li>
<li>Make <code>numeric</code> columns apply a gradient as if a categorical color were supplied. Basic idea is that instead of assigning categorical colors to each numeric value, for numeric columns assign the categorical color to the column name. Then it should proceed as it does now, when the user supplies a <code>color_sub</code> for that column name.</li>
</ul></li>
</ul></div>
<div class="section level1">
<h1 id="todo-05apr2022">TODO 05apr2022<a class="anchor" aria-label="anchor" href="#todo-05apr2022"></a></h1>
<div class="section level2">
<h2 id="salmon-import-to-summarizedexperiment">Salmon import to SummarizedExperiment?<a class="anchor" aria-label="anchor" href="#salmon-import-to-summarizedexperiment"></a></h2>
<ul><li>
<p>goal would be to mimic <code><a href="reference/import_proteomics_PD.html">import_proteomics_PD()</a></code> by importing <code>quant.sf</code> files, incorporate <code>curation.txt</code> for sample annotations.</p>
<ul><li>requires <code>GTF</code> and/or <code>tx2gene</code> file for transcript-to-gene.</li>
<li>Produce <code>list</code> with <code>TxSE</code> and <code>GeneSE</code> objects.</li>
<li>
<code>assays()</code> will include <code>counts</code> and <code>abundance</code>.</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="salmon-metadata">Salmon metadata<a class="anchor" aria-label="anchor" href="#salmon-metadata"></a></h2>
<p>COMPLETE: Salmon produces a useful file <code>"flenDist.txt"</code> that includes the distribution of fragment lengths observed, from 1 to 1000 length. Goal is to parse this file to determine the weighted mean fragment length, likely extending <code><a href="reference/get_salmon_meta.html">get_salmon_meta()</a></code>.</p>
</div>
</div>
<div class="section level1">
<h1 id="todo-21mar2022">TODO 21mar2022<a class="anchor" aria-label="anchor" href="#todo-21mar2022"></a></h1>
<div class="section level2">
<h2 id="ucsc-track-hubs">UCSC track hubs<a class="anchor" aria-label="anchor" href="#ucsc-track-hubs"></a></h2>
<ul><li>
<p><code><a href="reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> updates:</p>
<ul><li>
<p>currently the compositeTrack output requires editing to become visible by default. It appears now to export a mix of superTrack/compositeTrack. The apparent changes required:</p>
<ol style="list-style-type: decimal"><li>Remove “compositeTrack on” from the 2nd level of compositeTrack view.</li>
<li>Remove “superTrack on show” from the 1st level of compositeTrack view.</li>
<li>Add “type bigWig” to the 1st level of compositeTrack view (unlikely to be necessary however).</li>
<li>Add to 3rd level compositeTrack view, for each track: <code>"subGroups            view=COV"</code> Unclear if this line is required for visibility.</li>
</ol></li>
<li>
<p>Option <code>autoScale group</code> is useful for sets of tracks, and it should be easy to configure upfront.</p>
<ul><li>Track groups should be possible to be created independent of track order. For example ATAC_cov and ATAC_NDR could appear for each sample in order, with ATAC_cov autoScale to its own group, and ATAC_NDR autoScale to its own group.</li>
</ul></li>
<li><p>Not all track arguments are being honored in the final track hub entry. For example autoScale=group is not propagating through the workflow.</p></li>
</ul></li>
</ul></div>
</div>
<div class="section level1">
<h1 id="todo-for-platjam">TODO for platjam<a class="anchor" aria-label="anchor" href="#todo-for-platjam"></a></h1>
<div class="section level2">
<h2 id="nmatlist2heatmaps"><code>nmatlist2heatmaps()</code><a class="anchor" aria-label="anchor" href="#nmatlist2heatmaps"></a></h2>
<ul><li>when <code>transform</code> is used, alter the y-axis numeric labels, and color legend numeric labels accordingly.</li>
<li>Optionally draw y-axis horizontal abline at y=0 when data contains positive/negative values.</li>
</ul></div>
<div class="section level2">
<h2 id="ucsc-track-hub">UCSC track hub<a class="anchor" aria-label="anchor" href="#ucsc-track-hub"></a></h2>
<ul><li>
<p>Add example for <code><a href="reference/parse_ucsc_gokey.html">parse_ucsc_gokey()</a></code> showing three kinds of tracks as input</p>
<ol style="list-style-type: decimal"><li>bigWig with pos/neg or F/R or some track grouping -&gt; multiWig overlay</li>
<li>bigWig without grouping -&gt; composite tracks</li>
<li>bigBed -&gt; composite Tracks</li>
</ol></li>
<li>
<p>COMPLETE:new arguments for bigBed tracks</p>
<ul><li>scoreFilter=0, scoreFilterLimits=“0:1000”</li>
<li>minGrayLevel=4</li>
<li>spectrum=“off”, scoreMin=0, scoreMax=1000</li>
</ul></li>
</ul></div>
</div>
<div class="section level1">
<h1 id="todo-for-platjam-28aug2020">TODO for platjam 28aug2020<a class="anchor" aria-label="anchor" href="#todo-for-platjam-28aug2020"></a></h1>
<div class="section level2">
<h2 id="specific-usability">Specific usability<a class="anchor" aria-label="anchor" href="#specific-usability"></a></h2>
<ul><li>
<p>For <code>panel_groups</code>:</p>
<ul><li>COMPLETE: when all heatmaps in a set of panels share the same color gradient, the color legend should be displayed once per group, and should be labeled using the <code>panel_groups</code> value.</li>
<li>make the <code>ht_gap</code> between adjacent heatmaps in the same group smaller</li>
<li>COMPLETE: hide y-axis labels on the profile plot for all but the last panel in an adjacent set. For example using <code>EnrichedHeatmap::anno_enriched(... yaxis=FALSE)</code>
</li>
</ul></li>
</ul></div>
<div class="section level2">
<h2 id="visual-enhancements">Visual enhancements<a class="anchor" aria-label="anchor" href="#visual-enhancements"></a></h2>
<ul><li>
<code>anno_df</code> should probably use <code>Heatmap()</code> and not <code>HeatmapAnnotation()</code> so that it can enable <code>use_raster=TRUE</code>.</li>
<li>Color legend for the profile plot should use lines instead of boxes to indicate colors (if possible).</li>
</ul></div>
<div class="section level2">
<h2 id="big-picture---make-it-easy-to-use">Big picture - make it easy to use<a class="anchor" aria-label="anchor" href="#big-picture---make-it-easy-to-use"></a></h2>
<p>Make coverage heatmaps easier for other scientists.</p>
<ul><li>
<p>tab-delimited table input with the following columns:</p>
<ul><li>coverage matrix file path - full path to coverage file</li>
<li>name - name associated to the coverage matrix file, used below</li>
<li>display - logical indicating whether to display this coverage matrix; TRUE, 1 means “yes”; FALSE, 0 means “no”. When “no”, the coverage may still be usable in difference matrix calculations.</li>
<li>label - need some way to specify line breaks, without using line breaks.</li>
<li>xaxis_names - three comma-delimited values, or <code>NA</code>, for example: <code>-250kb,center,+250kb</code>
</li>
<li>group - used for <code>panel_groups</code>
</li>
<li>signal_ceiling</li>
<li>ylim</li>
<li>color_gradient - string that matches jam_linear, jam_divergent, viridis, RColorBrewer, or a comma-delimited series of hex colors</li>
<li>control_name - optional name of a control coverage file to be used to create a difference matrix</li>
</ul></li>
<li><p>tab-delimited anno_df file whose first column values should match rownames in each coverage matrix file.</p></li>
</ul><p>Make jam color gradients available:</p>
<ul><li>
<code>jam_linear</code> - color blind friendly linear gradients from white to color</li>
<li>
<code>jam_divergent</code> - color blind friendly divergent gradients with black as the central color</li>
<li>The gradients above are designed so that the divergent start or end color matches the color name of linear gradients. The intent is to use linear gradients to represent straight coverage, and divergent gradients for coverage differences.</li>
</ul><p>Make it possible to define coverage differences</p>
<ul><li>Probably use some format in the tab-delimited file above.</li>
</ul></div>
<div class="section level2">
<h2 id="k-means-and-partitioning-correct-the-order">k-means and partitioning, correct the order<a class="anchor" aria-label="anchor" href="#k-means-and-partitioning-correct-the-order"></a></h2>
<p>Currently k-means clustering and partitioning works together, but the k-means is performed on everything, then clusters are divided by partition.</p>
<p>The effect when partitioning by “up” and “down” is that some k-means clusters are dominated by the “up” and “down” such that cluster 1 has 95% members from “up”, very few from down. For k=4, it ends up making 8 clusters, which are mostly just 4 proper clusters, and 4 junk clusters (the “up” cluster with few “down” members).</p>
<p>Would be better to k-means within each partition, to make proper subclusters for each partition.</p>
</div>
</div>
<div class="section level1">
<h1 id="todo-for-platjam-22may2020">TODO for platjam 22may2020<a class="anchor" aria-label="anchor" href="#todo-for-platjam-22may2020"></a></h1>
<div class="section level2">
<h2 id="dynamic-ylims-done">Dynamic ylims (DONE)<a class="anchor" aria-label="anchor" href="#dynamic-ylims-done"></a></h2>
<ul><li>The detected ylim range should include the additional range when show_error=TRUE. This change involves calculating the error bars during the ylim calculations, not too computationally intensive, but it must replicate the method used by EnrichedHeatmap.</li>
</ul></div>
<div class="section level2">
<h2 id="custom-color-substition-colorsub-done">Custom color substition (colorSub) (DONE)<a class="anchor" aria-label="anchor" href="#custom-color-substition-colorsub-done"></a></h2>
<ul><li>
<code><a href="reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> should allow optional <code>colorSub</code> argument as a named vector of colors. Any categorical column all of whose values match <code>names(colorSub)</code> will use colors in <code>colorSub</code> instead of generating its own new categorical colors. This option will help allow colors to be consistent.</li>
</ul></div>
<div class="section level2">
<h2 id="allow-both-partition-and-kmeans-clustering-done">Allow both partition and kmeans clustering (DONE)<a class="anchor" aria-label="anchor" href="#allow-both-partition-and-kmeans-clustering-done"></a></h2>
<ul><li>Currently kmeans clustering overrides partition, but we want the option to cluster and partition.</li>
</ul></div>
<div class="section level2">
<h2 id="genome-coverage-matrix-methods-done">Genome coverage matrix methods (DONE)<a class="anchor" aria-label="anchor" href="#genome-coverage-matrix-methods-done"></a></h2>
<ul><li>
<code><a href="reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> argument for <code>data.frame</code> of annotations, used to display alongside heatmaps, and/or used to sort the heatmap rows. For example one column could contain <code>"log2foldchange"</code>, or a categorical variable.</li>
<li>add a new file importer for deepTools coverage matrix files. Allow for multiple samples to be present in the same file, then create one <code>normalizedMatrix</code> for each sample with consistent rownames and colnames.</li>
</ul></div>
<div class="section level2">
<h2 id="genome-coverage-profile-methods">Genome coverage profile methods<a class="anchor" aria-label="anchor" href="#genome-coverage-profile-methods"></a></h2>
<ul><li>Similar to <code><a href="reference/nmatlist2heatmaps.html">nmatlist2heatmaps()</a></code> except that it focuses on just the profile plots, including optional error bars and statistical testing by position.</li>
</ul></div>
<div class="section level2">
<h2 id="custom-x-axis-range">Custom x-axis range<a class="anchor" aria-label="anchor" href="#custom-x-axis-range"></a></h2>
<ul><li>Ability to specify a custom subset range for the x-axis. Note that matrices in <code>nmatlist</code> are not required to share the same x-axis range, so this function would probably need to be applied to individual <code>normalizedMatrix</code> entries.</li>
</ul></div>
</div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

