<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Import Salmon quant.sf files to SummarizedExperiment — import_salmon_quant • platjam</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Import Salmon quant.sf files to SummarizedExperiment — import_salmon_quant"><meta property="og:description" content="Import Salmon quant.sf files to SummarizedExperiment"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">platjam</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.73.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="http://github.com/jmw86069/platjam/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import Salmon quant.sf files to SummarizedExperiment</h1>
    <small class="dont-index">Source: <a href="http://github.com/jmw86069/platjam/blob/HEAD/R/platjam-import-salmonquant.R" class="external-link"><code>R/platjam-import-salmonquant.R</code></a></small>
    <div class="hidden name"><code>import_salmon_quant.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Import Salmon quant.sf files to SummarizedExperiment</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">import_salmon_quant</span><span class="op">(</span></span>
<span>  <span class="va">salmonOut_paths</span>,</span>
<span>  import_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tx"</span>, <span class="st">"gene"</span>, <span class="st">"gene_body"</span>, <span class="st">"gene_tx"</span><span class="op">)</span>,</span>
<span>  gtf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tx2gene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  curation_txt <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tx_colname <span class="op">=</span> <span class="st">"transcript_id"</span>,</span>
<span>  gene_colname <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>  gene_body_colname <span class="op">=</span> <span class="st">"transcript_type"</span>,</span>
<span>  geneFeatureType <span class="op">=</span> <span class="st">"exon"</span>,</span>
<span>  txFeatureType <span class="op">=</span> <span class="st">"exon"</span>,</span>
<span>  countsFromAbundance <span class="op">=</span> <span class="st">"lengthScaledTPM"</span>,</span>
<span>  gene_body_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trim_tx_from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"[(][-+][)]"</span><span class="op">)</span>,</span>
<span>  trim_tx_to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>salmonOut_paths</dt>
<dd><p><code>character</code> vectors to each individual folder
that contains the <code>"quant.sf"</code> output file for Salmon.</p></dd>


<dt>import_types</dt>
<dd><p><code>character</code> indicating which type or types of
data to return. Note that the distinction between <code>gene</code> and
<code>gene_body</code> is only relevant when there are transcript entries
defined with <code>transcript_type="gene_body"</code>. These entries specifically
represent unspliced transcribed regions for a gene locus, and
only for multi-exon genes.</p><ul><li><p><code>tx</code>: transcript quantitation, direct import of <code>quant.sf</code> files.</p></li>
<li><p><code>gene</code>: gene quantitation after calling <code><a href="https://rdrr.io/pkg/tximport/man/summarizeToGene.html" class="external-link">tximport::summarizeToGene()</a></code>,
excluding <code>transcript_type="gene_body"</code>.</p></li>
<li><p><code>gene_body</code>: gene quantitation after calling <code><a href="https://rdrr.io/pkg/tximport/man/summarizeToGene.html" class="external-link">tximport::summarizeToGene()</a></code>,
including <code>transcript_type="gene_body"</code>.</p></li>
</ul></dd>


<dt>gtf</dt>
<dd><p><code>character</code> path to a GTF file, used only when <code>tx2gene</code>
is not supplied. When used, <code><a href="https://rdrr.io/pkg/splicejam/man/makeTx2geneFromGtf.html" class="external-link">splicejam::makeTx2geneFromGtf()</a></code> is
called to create a <code>data.frame</code> object <code>tx2gene</code>.</p></dd>


<dt>tx2gene</dt>
<dd><p><code>character</code> path to file, or <code>data.frame</code> with at
least two columns matching <code>tx_colname</code> and <code>gene_colname</code> below.
When supplied, the <code>gtf</code> argument is ignored, unless the file
path is not accessible, or the data is not <code>data.frame</code>.</p></dd>


<dt>curation_txt</dt>
<dd><p><code>data.frame</code> whose first column should match the
sample column headers found in the PD abundance columns, and
subsequent columns contain associated sample annotations.
If <code>curation_txt</code> is not supplied, then values will be split into
columns by <code>_</code> underscore or <code>" "</code> whitespace characters.</p></dd>


<dt>tx_colname, gene_colname</dt>
<dd><p><code>character</code> strings indicating colnames
in <code>tx2gene</code> that should be used.</p><ul><li><p><code>tx_colname</code> represents unique identifier for each transcript,
usually <code>"transcript_id"</code>.</p></li>
<li><p><code>gene_colname</code> represents a gene label associated with gene
summarized expression values, typically <code>"gene_name"</code>.</p></li>
</ul></dd>


<dt>geneFeatureType, txFeatureType</dt>
<dd><p><code>character</code> arguments passed to
<code><a href="https://rdrr.io/pkg/splicejam/man/makeTx2geneFromGtf.html" class="external-link">splicejam::makeTx2geneFromGtf()</a></code> only when supplying argument
<code>gtf</code> with a path to a GTF file.</p></dd>


<dt>countsFromAbundance</dt>
<dd><p><code>character</code> string passed to
<code><a href="https://rdrr.io/pkg/tximport/man/summarizeToGene.html" class="external-link">tximport::summarizeToGene()</a></code> to define the method for calculating
abundance.</p></dd>


<dt>gene_body_ids</dt>
<dd><p><code>character</code> optional vector with specific row
identifiers that should be considered <code>transcript_type="gene_body"</code>
entries, relevant to argument <code>import_types</code> above. When <code>gene_body_ids</code>
is defined, these entries are used directly without using <code>tx2gene</code>.
When <code>gene_body_ids</code> is not defined, <code>tx2gene$transcript_type</code> is used
if present. If that column is not present, or does not contain any
entries with <code>"gene_body"</code>, then all transcripts are used for
<code>import_types="gene"</code>, and <code>import_types="gene_body"</code> is not valid
and therefore is not returned.</p></dd>


<dt>verbose</dt>
<dd><p><code>logical</code> indicating whether to print verbose output.</p></dd>


<dt>...</dt>
<dd><p>additional arguments are passed to supporting functions.</p></dd>


<dt>curate_tx_from, curate_tx_to</dt>
<dd><p><code>character</code> vector of regular expression
patterns to be used optionally to curate the values in <code>tx_colname</code> prior
to joining those values to <code>tx2gene[[tx_colname]]</code>.
The default is to remove <code>"(-)"</code> and <code>"(+)"</code> from the transcript_id
(<code>tx_colname</code>) column.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p><code>list</code> with <code>SummarizedExperiment</code> objects, each of which
contain assay names <code>c("counts", "abundance", "length)</code>, where
<code>c("counts", "abundance")</code> are transformed with <code>log2(1 + x)</code>.
The transform can be reversed with <code>10^x - 1</code>.
The <code>SummarizedExperiment</code> objects by name:</p><ul><li><p><code>"TxSE"</code>: transcript-level values imported from <code>quant.sf</code>.</p></li>
<li><p><code>"GeneSE"</code>: gene-level summary values, excluding
<code>"gene_body"</code> entries.</p></li>
<li><p><code>"GeneBodySE"</code>: gene-level summary values, including
<code>"gene_body"</code> entries.</p></li>
<li><p><code>"GeneTxSE"</code>: gene-level summary values, where transcripts are
combined to gene level, and <code>"gene_body"</code> entries are represented
separately, with suffix <code>"_gene_body"</code> added to the gene name.</p></li>
</ul></div>
    <div id="details">
    <h2>Details</h2>
    <p>This function is intended to automate the process of importing
a series of <code>quant.sf</code> files, then generating <code>SummarizedExperiment</code>
objects at the transcript and gene level. It optionally includes
sample annotation provided as a <code>data.frame</code> in argument <code>curation_txt</code>.
It also includes transcript and gene annotations through either
<code>data.frame</code> from argument <code>tx2gene</code>, or it derives <code>tx2gene</code>
from a GTF file from argument <code>gtf</code>. The GTF file option then calls
<code><a href="https://rdrr.io/pkg/splicejam/man/makeTx2geneFromGtf.html" class="external-link">splicejam::makeTx2geneFromGtf()</a></code>.</p>
<p>This function can optionally process data that includes full length
gene body regions, annotated with <code>"gene_body"</code>. This option is specific
for Salmon quantitation where the transcripts include full length
gene body for multi-exon genes, for example to measure unspliced
transcript abundance.</p><ul><li><p><code>import_types="gene"</code> summarizes only the proper transcripts,
excluding <code>"gene_body"</code> entries.</p></li>
<li><p><code>import_types="gene_body"</code> summarizes all transcript
and full gene entries into one summary transcript abundance.</p></li>
<li><p><code>import_types="gene_tx"</code> summarizes proper transcript to gene level,
and separately represents <code>"gene_body"</code> entries for comparison.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p>Other jam import functions: 
<code><a href="coverage_matrix2nmat.html">coverage_matrix2nmat</a>()</code>,
<code><a href="deepTools_matrix2nmat.html">deepTools_matrix2nmat</a>()</code>,
<code><a href="frequency_matrix2nmat.html">frequency_matrix2nmat</a>()</code>,
<code><a href="import_lipotype_csv.html">import_lipotype_csv</a>()</code>,
<code><a href="import_metabolomics_niehs.html">import_metabolomics_niehs</a>()</code>,
<code><a href="import_nanostring_csv.html">import_nanostring_csv</a>()</code>,
<code><a href="import_nanostring_rcc.html">import_nanostring_rcc</a>()</code>,
<code><a href="import_nanostring_rlf.html">import_nanostring_rlf</a>()</code>,
<code><a href="import_proteomics_PD.html">import_proteomics_PD</a>()</code>,
<code><a href="import_proteomics_mascot.html">import_proteomics_mascot</a>()</code>,
<code><a href="process_metab_compounds_file.html">process_metab_compounds_file</a>()</code></p></div>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by James M. Ward.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

